query/common/src/main/java/org/codice/ddf/admin/common/PrioritizedBatchExecutor.java

```diff
@@ -0,0 +1,242 @@
+     * @param threadPoolSize size of the underlying {@code ExecutorService}. 1-64 size is valid. If an
+     *                       argument higher than 64 is detected, it will default to the max number of threads.
+     * @param tasks          a non-null {@code List} of tasks that will be executed in order
+     * @param taskHandler    a non-null task handler that determines if a task result is valid to return
+     */
+    public PrioritizedBatchExecutor(int threadPoolSize, List<List<Callable<T>>> tasks,
+            Function<T, T> taskHandler) {
+        if (tasks == null) {
+            throw new IllegalArgumentException("Argument {tasks} cannot be null.");
+        }
+
+        if (taskHandler == null) {
+            throw new IllegalArgumentException("Argument {taskHandler} cannot be null.");
+        }
+
+        if (threadPoolSize > MAX_THREAD_POOL_SIZE) {
+            LOGGER.debug(String.format(
+                    "Thread pool size exceeds maximum allowed value of %d, defaulting to number of processors available to the JVM [{}].",
+                    MAX_THREAD_POOL_SIZE));
+
+            threadPoolSize = MAX_THREAD_POOL_SIZE;
+        }
+
+        this.tasks = tasks;
+        this.taskHandler = taskHandler;
+
+        threadPool = Executors.newFixedThreadPool(threadPoolSize);
+    }
+
+    /**
+     * Start task execution and blocks until the highest priority task batch has returned a valid result
+     * according to the task handler, then cleans up remaining tasks. The current instance of the
+     * {@code PrioritizedBatchExecutor} is not useable after calling {@code getFirst()}.
+     *
+     * @return an {@code Optional} containing a task's result, if there was one
+     */
+    public Optional<T> getFirst() {
+        try {
+            List<CompletionService<T>> prioritizedCompletionServices =
+                    getPrioritizedCompletionServices();
+
+            for (int i = 0; i < tasks.size(); i++) {
+                List<Callable<T>> taskBatch = tasks.get(i);
+                CompletionService<T> completionService = prioritizedCompletionServices.get(i);
+
+                LOGGER.debug("Executing batch {} of {}.", i + 1, tasks.size());
+                for (int j = 0; j < taskBatch.size(); j++) {
+                    try {
+                        LOGGER.debug("\tAttempting to take from completion service queue.");
+                        Future<T> taskFuture = completionService.take();
+
+                        Optional<T> result = handleTaskResult(taskFuture);
+                        if (result.isPresent()) {
+                            LOGGER.debug("\tReturning valid task result {} of {} tasks.",
+                                    j + 1,
+                                    taskBatch.size());
+                            return result;
+                        }
+                    } catch (InterruptedException e) {
+                        LOGGER.debug("\t\tFailed to get future from completion service.");
```

Comment:
✏️ You really are going to want to do something with your (unlikely) `InterruptedException`. Your choices here are to interrupt the current thread or propagate the exception. 
`Thread.currentThread().interrupt();` will give you the former and be the simplest/cleanest solution.

Refer to [this article from Brian Goetz](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html) for more details.